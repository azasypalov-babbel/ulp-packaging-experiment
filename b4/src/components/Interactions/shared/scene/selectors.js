export const getItem = (state) => state.item;

export const isGap = (node) => node.type === 'gap';

export const getGaps = (item) => item.nodes.filter(isGap);


export const getActiveNode = (state) => {
  const item = getItem(state);

  if (!item || state.completed) {
    return null;
  }

  const { activeNodeIndex } = state;
  const activeNode = item.nodes[activeNodeIndex];
  if (activeNode && isGap(activeNode)) {
    return activeNode;
  }
  return null;
};

export const nextGapNodeIndex = (item, minIndex = 0) => {
  return item.nodes.findIndex((node, index) => {
    return isGap(node) && index >= minIndex;
  });
};

export const hasGaps = (item) => {
  return nextGapNodeIndex(item, 0) !== -1;
};

export const hasGapAfter = (item, minIndex = 0) => {
  return nextGapNodeIndex(item, minIndex) !== -1;
};

export const getMistakeCount = (state) => {
  const activeItem = getItem(state);
  return activeItem.nodes
    .filter(isGap)
    .filter((node) => !node.attempt.pending)
    .reduce((acc, { attempt: { number, solved } }) => {
      return solved ? acc + (number - 1) : acc + number;
    }, 0);
};
