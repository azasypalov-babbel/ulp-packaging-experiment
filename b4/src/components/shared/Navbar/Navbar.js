import React from 'react';
import PropTypes from 'prop-types';
import cx from '../../../lib/cxHelper';
import zendeskWidget from '../../../lib/zendeskWidget';
import ProgressBar, { progressBarPropTypes } from './ProgressBar';
import ProgressCounter, { progressCounterPropTypes } from './ProgressCounter';
import SkipTrainerControls from '../../Sequence/SkipTrainerControls';
import CrossIcon from '../icons/CrossIcon';
import withTranslations from '../withTranslations';

const CSS_NAME = cx('cascada-navbar');

export const Navbar = ({
  onClickClose,
  progressCounter,
  progressBar,
  translations,
  showFeedbackButton,
  showProgressBar
}) => {
  const renderCloseBtn = () => (
    <a
      tabIndex="0"
      onClick={onClickClose}
      data-selector="navbar-close-link"
    >
      { translations.closeButtonTitle }
    </a>
  );

  const renderFeedbackBtn = () => showFeedbackButton && (
    <a
      tabIndex="0"
      className={`${CSS_NAME}__feedback`}
      onClick={() => {
        zendeskWidget.toggle();
      }}
    >
      { translations.feedbackButtonTitle }
    </a>
  );

  const renderCloseIcon = () => (
    <a
      tabIndex="0"
      className={`${CSS_NAME}__close-icon`}
      onClick={onClickClose}
      data-selector="navbar-close-icon"
    >
      <CrossIcon size={'1.8rem'} fillColor={'currentColor'}/>
    </a>
  );

  return  (
    <header className={CSS_NAME}>
      <div className={`${CSS_NAME}__gate`}>
        <div className={`${CSS_NAME}__item ${CSS_NAME}__item--medium-only`}>
          <span className={`${CSS_NAME}__brand`}>+B</span>
        </div>
        <SkipTrainerControls renderSkipAreas={showProgressBar}>
          <div className={`${CSS_NAME}__item ${CSS_NAME}__item--progress`}>
            {showProgressBar && <ProgressBar
              {...progressBar}
            />}
            {showProgressBar && <ProgressCounter
              {...progressCounter}
            />}
          </div>
        </SkipTrainerControls>
        <div className={`${CSS_NAME}__item ${CSS_NAME}__item--medium-only ${CSS_NAME}__item--align-right`}>
          {renderFeedbackBtn()}
          {renderCloseBtn()}
        </div>
        <div className={`${CSS_NAME}__item ${CSS_NAME}__item--small-only ${CSS_NAME}__item--align-left`}>
          {renderCloseIcon()}
        </div>
      </div>
    </header>
  );
};

Navbar.propTypes = {
  onClickClose: PropTypes.func.isRequired,
  progressBar: PropTypes.shape(progressBarPropTypes).isRequired,
  progressCounter: PropTypes.shape(progressCounterPropTypes).isRequired,
  showFeedbackButton: PropTypes.bool.isRequired,
  showProgressBar: PropTypes.bool,
  translations: PropTypes.shape({
    closeButtonTitle: PropTypes.string.isRequired,
    feedbackButtonTitle: PropTypes.string.isRequired
  }).isRequired
};

const getTranslations = (translate) => ({
  closeButtonTitle: translate('navbar.close'),
  feedbackButtonTitle: translate('navbar.contact')
});

export default withTranslations(getTranslations)(Navbar);
