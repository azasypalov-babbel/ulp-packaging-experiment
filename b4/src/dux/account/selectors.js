import { createSelector } from 'reselect';

const accountSelector = (state) => state.data;

export const displayName = createSelector(
  accountSelector,
  (account) => account.displayname
);

export const email = createSelector(
  accountSelector,
  (account) => account.email
);

// Introduced for smart-surfaces feature
export const createdAt = createSelector(
  accountSelector,
  (account) => account.createdAt
);

export const isBabbelUser = createSelector(
  accountSelector,
  (account) => {
    const domain = account.email?.split('@')[1] || '';
    return domain === 'babbel.com';
  }
);
