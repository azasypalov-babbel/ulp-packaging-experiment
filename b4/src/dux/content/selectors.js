import { createSelector } from 'reselect';
import { flatten } from 'underscore';

const selectData = (state) => state.data;
const selectCourseOverviews = createSelector(
  selectData,
  (data) => data ? data.courseOverviews : null
);

export const isUnlocked = createSelector(
  selectData,
  (data) => data ? data.unlocked : null
);

export const lessons = createSelector(
  selectCourseOverviews,
  (courseOverviews) => {
    if (!courseOverviews) return null;

    return flatten(courseOverviews.map(({ courses }) =>
      courses.map(({ lessons }) => lessons)));
  });

// returns slice of content with only courseOverview, course and lesson
// related to given learningActivityId
export const contentSliceOfCurrentLesson = (state, learningActivityId) => {
  const courseOverviews = selectCourseOverviews(state);

  if (!courseOverviews) return null;

  let slice = [];

  for (let i = 0; i < courseOverviews.length; i += 1) {
    const course = courseOverviews[i].courses.find(({ lessons }) =>
      Boolean(lessons.find(({ id }) => id === learningActivityId)));

    if (course) {
      slice.push({ ...courseOverviews[i], courses: [course] });
      break;
    }
  }

  return {
    courseOverviews: slice
  };
};

export const currentCourseOverview = (state, learningActivityId) => {
  const contentSlice = contentSliceOfCurrentLesson(state, learningActivityId);

  if (!contentSlice || !contentSlice.courseOverviews) {
    return null;
  }

  return contentSlice.courseOverviews[0];
};

export const currentCourse = (state, learningActivityId) => {
  const courseOverview = currentCourseOverview(state, learningActivityId);

  if (!courseOverview) return null;

  return courseOverview.courses[0];
};

export const currentLesson = (state, learningActivityId) => {
  const courseLessons = lessons(state);

  if (!courseLessons) {
    return null;
  }

  return courseLessons.find(({ id }) => id === learningActivityId) || null;
};
