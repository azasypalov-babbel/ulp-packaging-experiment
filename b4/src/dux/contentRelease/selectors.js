export const contentReleaseId = (state) => {
  if (!state.data) return null;

  return state.data.id;
};
