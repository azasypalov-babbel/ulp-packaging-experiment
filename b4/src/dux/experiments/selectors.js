import camelize from 'camelize';

// current experiments targets
export const ACTIVE_EXPERIMENTS = {};

export const getExperimentBucket = (target, state) => {
  return state[camelize(target)];
};

