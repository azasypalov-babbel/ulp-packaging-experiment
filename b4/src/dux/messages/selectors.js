
export const getFirstMessage = (state) => state[0];

