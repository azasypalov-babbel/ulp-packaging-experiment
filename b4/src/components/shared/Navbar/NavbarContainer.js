import React, { Component } from 'react';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import * as sequenceSelectors from '../../../dux/sequence/selectors';
import * as sessionSelectors from '../../../dux/session/selectors';
import * as reviewActions from '../../../dux/review/actions';
import * as lessonActions from '../../../dux/lesson/actions';
import * as sequenceActions from '../../../dux/sequence/actions';
import { track } from '../../../dux/tracker/actions';
import Navbar from './Navbar';
import { progressCounterPropTypes } from './ProgressCounter';

export const mapStateToProps = ({ zendeskWidget, session, sequence }) => {
  const isLesson = sessionSelectors.isLesson(session);
  const progressCounter = isLesson
    ? sequenceSelectors.sequenceProgressCounter(sequence)
    : sequenceSelectors.itemProgressCounter(sequence);

  return {
    isLessonLandingScreenShown: sessionSelectors.isLessonLandingScreenShown(session),
    isLesson,
    progressBar: {
      trainerCount: sequenceSelectors.trainersInSequence(sequence).length,
      currentTrainerIndex: sequence.currentTrainerIndex,
      sequenceHeadProgress: sequenceSelectors.sequenceHeadProgress(sequence),
      sequenceHeadIndex: sequenceSelectors.sequenceHeadIndex(sequence)
    },
    showProgressBar: sequence.started,
    showFeedbackButton: zendeskWidget.ready,
    progressCounter
  };
};

const mapDispatchToProps = {
  closeLesson: lessonActions.closeLesson,
  closeReview: reviewActions.closeReview,
  track,
  navigate: sequenceActions.navigate
};

export class NavbarContainer extends Component {
  constructor(props) {
    super(props);
    this.handleProgressBarTrainerClick = this.handleProgressBarTrainerClick.bind(this);
    this.handleCloseClick = this.handleCloseClick.bind(this);
  }

  handleCloseClick() {
    const { isLesson, closeLesson, closeReview, isLessonLandingScreenShown, track } = this.props;

    if (isLessonLandingScreenShown) {
      /* eslint-disable camelcase */
      track({
        event: 'gui:interacted',
        version: 1,
        payload: {
          gui_element: 'close_button',
          gui_container: 'lesson_landing_page',
          origin: 'lesson_player',
          interaction: 'clicked'
        }
      });
      /* eslint-enable camelcase */
    }

    if (isLesson) {
      closeLesson();
    } else {
      closeReview();
    }
  }

  handleProgressBarTrainerClick() {
    const { track, isLesson } = this.props;

    if (isLesson) {
      /* eslint-disable camelcase */
      track({
        event: 'gui:interacted',
        version: 1,
        payload: {
          gui_element: 'progress_bar',
          origin: 'lesson_player',
          interaction: 'clicked'
        }
      });
      /* eslint-enable camelcase */
    }
  }

  render() {
    const {
      progressBar,
      progressCounter,
      showFeedbackButton,
      showProgressBar
    } = this.props;

    return (
      <div className="loy">
        <Navbar
          onClickClose={this.handleCloseClick}
          showFeedbackButton={showFeedbackButton}
          showProgressBar={showProgressBar}
          progressBar={{
            ...progressBar,
            onTrainerClick: this.handleProgressBarTrainerClick
          }}
          progressCounter={progressCounter}
        />
      </div>
    );
  }
}

NavbarContainer.propTypes = {
  track: PropTypes.func.isRequired,
  closeLesson: PropTypes.func.isRequired,
  closeReview: PropTypes.func.isRequired,
  progressBar: PropTypes.object.isRequired,
  progressCounter: PropTypes.shape(progressCounterPropTypes).isRequired,
  showFeedbackButton: PropTypes.bool.isRequired,
  showProgressBar: PropTypes.bool,
  isLesson: PropTypes.bool.isRequired,
  isLessonLandingScreenShown: PropTypes.bool.isRequired
};

export default compose(
  connect(mapStateToProps, mapDispatchToProps)
)(NavbarContainer);
