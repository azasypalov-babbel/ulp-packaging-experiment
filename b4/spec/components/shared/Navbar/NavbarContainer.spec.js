import React from 'react';
import { NavbarContainer, mapStateToProps } from '../../../../src/components/shared/Navbar/NavbarContainer';
import * as sequenceSelectors from '../../../../src/dux/sequence/selectors';
import * as sessionSelectors from '../../../../src/dux/session/selectors';
import { shallow } from 'enzyme';


jest.mock('../../../../src/dux/sequence/selectors');
jest.mock('../../../../src/dux/session/selectors', () => ({
  isLesson: jest.fn(),
  isLessonLandingScreenShown: jest.fn()
}));

const defaultProps = {
  isLessonLandingScreenShown: false,
  progressBar: {
    trainerCount: 4,
    sequenceHeadIndex: 2,
    currentTrainerIndex: 2,
    sequenceHeadProgress: 0.33,
    onTrainerClick: jest.fn()
  },
  showFeedbackButton: true,
  showProgressBar: true,
  progressCounter: {
    current: 2,
    total: 3
  },
  itemProgressCounter: {
    current: 5,
    total: 12
  },
  isLesson: true,
  closeLesson: jest.fn(),
  closeReview: jest.fn(),
  track: jest.fn()
};

describe('<NavbarContainer />', () => {
  describe('mapStateToProps', () => {
    beforeEach(() => {
      sequenceSelectors.trainersInSequence.mockImplementation(() => []);
    });

    const state = {
      zendeskWidget: {},
      sequence: {},
      session: {
        lessonLandingScreenSettings: {}
      }
    };

    test('selects data from the state', () => {
      mapStateToProps(state);

      expect(sequenceSelectors.trainersInSequence).toHaveBeenCalled();
      expect(sequenceSelectors.sequenceHeadProgress).toHaveBeenCalled();
      expect(sequenceSelectors.sequenceHeadIndex).toHaveBeenCalled();
      expect(sessionSelectors.isLesson).toHaveBeenCalled();
      expect(sessionSelectors.isLessonLandingScreenShown).toHaveBeenCalled();
    });

    describe('in lesson', () => {
      beforeEach(() => {
        sessionSelectors.isLesson.mockImplementation(() => true);
      });

      afterEach(() => {
        sessionSelectors.isLesson.mockClear();
      });

      test('progressCounter is based on trainers', () => {
        mapStateToProps(state);

        expect(sequenceSelectors.sequenceProgressCounter).toHaveBeenCalled();
        expect(sequenceSelectors.itemProgressCounter).not.toHaveBeenCalled();
      });
    });

    describe('in review', () => {
      beforeEach(() => {
        sessionSelectors.isLesson.mockImplementation(() => false);
      });

      afterEach(() => {
        sessionSelectors.isLesson.mockClear();
      });

      test('progressCounter is based on items', () => {
        mapStateToProps(state);

        expect(sequenceSelectors.sequenceProgressCounter).not.toHaveBeenCalled();
        expect(sequenceSelectors.itemProgressCounter).toHaveBeenCalled();
      });
    });
  });

  describe('wraps Navbar component with all necessary data', () => {
    test('in lesson mode', () => {
      const component = shallow(<NavbarContainer isLesson={true} {...defaultProps} />);

      expect(component).toMatchSnapshot();
    });

    test('in review mode', () => {
      const component = shallow(<NavbarContainer isLesson={false} {...defaultProps} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe('handleCloseClick', () => {
    test('calls closeLesson in lesson mode', () => {
      const closeLessonMock = jest.fn();
      const component = shallow(<NavbarContainer
        {...defaultProps}
        isLesson={true}
        closeLesson={closeLessonMock}
      />);
      component
        .instance()
        .handleCloseClick();

      expect(closeLessonMock).toHaveBeenCalled();
    });

    test('calls closeReview in review mode', () => {
      const closeReviewMock = jest.fn();
      const component = shallow(<NavbarContainer
        {...defaultProps}
        isLesson={false}
        closeReview={closeReviewMock}
      />);
      component
        .instance()
        .handleCloseClick();

      expect(closeReviewMock).toHaveBeenCalled();
    });
  });

  describe('handleProgressBarTrainerClick', () => {
    beforeEach(() => {
      defaultProps.track.mockClear();
    });

    test('it calls track with required payload for lessons', () => {
      const component = shallow(<NavbarContainer {...defaultProps} isLesson={true}/>);
      component
        .instance()
        .handleProgressBarTrainerClick(0);

      expect(defaultProps.track).toHaveBeenCalled();
      expect(defaultProps.track.mock.calls[0][0]).toMatchSnapshot();
    });
  });
});
