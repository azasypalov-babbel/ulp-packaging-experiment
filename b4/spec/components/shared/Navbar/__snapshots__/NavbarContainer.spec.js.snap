// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavbarContainer /> handleProgressBarTrainerClick it calls track with required payload for lessons 1`] = `
Object {
  "event": "gui:interacted",
  "payload": Object {
    "gui_element": "progress_bar",
    "interaction": "clicked",
    "origin": "lesson_player",
  },
  "version": 1,
}
`;

exports[`<NavbarContainer /> wraps Navbar component with all necessary data in lesson mode 1`] = `
<div
  className="loy"
>
  <WithServices(WithTranslations(Navbar))
    onClickClose={[Function]}
    progressBar={
      Object {
        "currentTrainerIndex": 2,
        "onTrainerClick": [Function],
        "sequenceHeadIndex": 2,
        "sequenceHeadProgress": 0.33,
        "trainerCount": 4,
      }
    }
    progressCounter={
      Object {
        "current": 2,
        "total": 3,
      }
    }
    showFeedbackButton={true}
    showProgressBar={true}
  />
</div>
`;

exports[`<NavbarContainer /> wraps Navbar component with all necessary data in review mode 1`] = `
<div
  className="loy"
>
  <WithServices(WithTranslations(Navbar))
    onClickClose={[Function]}
    progressBar={
      Object {
        "currentTrainerIndex": 2,
        "onTrainerClick": [Function],
        "sequenceHeadIndex": 2,
        "sequenceHeadProgress": 0.33,
        "trainerCount": 4,
      }
    }
    progressCounter={
      Object {
        "current": 2,
        "total": 3,
      }
    }
    showFeedbackButton={true}
    showProgressBar={true}
  />
</div>
`;
